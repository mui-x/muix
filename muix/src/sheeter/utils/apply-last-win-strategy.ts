import { TAtomize } from 'reactxx-typings'
import { globalOptions } from '../global-state'

export interface ApplyLastWinStrategyResult {
    style?: TAtomize.AtomicWebsLow | TAtomize.AtomicNativeLow
    defferedIdxs?: number[]
}
export const enum AttemptType {
    first,
    firstIgnore,
    second
}

export type ApplyLastWinStrategyLow = (values: TAtomize.AtomicWeb[], attemptType: AttemptType) => ApplyLastWinStrategyResult

export const applyLastwinsStrategyRoot = (values: TAtomize.AtomicArray, applyLastWinStrategyLow: ApplyLastWinStrategyLow) => {
    const processDeffereds = globalOptions.processDeffereds
    let res = applyLastWinStrategyLow(values, processDeffereds ? AttemptType.first : AttemptType.firstIgnore)
    if (res.style) return res.style
    processDeffereds(values, res.defferedIdxs, values.state)
    res = applyLastWinStrategyLow(values, AttemptType.second)
    return res.style
}
