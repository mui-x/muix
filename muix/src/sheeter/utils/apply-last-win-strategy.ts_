import { TAtomize } from 'reactxx-typings'

// export interface ApplyLastWinStrategyResult {
//     style?: TAtomize.AtomicWebsLow | TAtomize.AtomicNativeLow
//     defferedIdxs?: number[]
// }
// export const enum AttemptType {
//     first,
//     firstIgnore,
//     second
// }

