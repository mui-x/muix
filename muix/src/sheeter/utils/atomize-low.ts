import { wrapPseudoPrefixes } from './wrap-pseudo-prefixes'
import { TTyped, TEngine } from 'reactxx-typings'

// platform dependent 
import { platform } from 'reactxx-sheeter'

//*******************************************************
//        adjustAtomizedLow

// linearize and atomize "ruleset"
// result is added to 'atomizedVariants'
export const adjustAtomizedLow = (
    ruleset: TEngine.Rulesets,
    atomizedVariants: TEngine.QueryableItems, path: string,
    pseudoPrefixes: string[], conditions: TEngine.Conditions
) => {
    processTree(ruleset, atomizedVariants, path, pseudoPrefixes, conditions)
}

export function isToAtomizeArray(obj): obj is TEngine.Ruleset[] { return obj && !(obj as TEngine.Queryables).$r$ && Array.isArray(obj) }
export function isAtomized(obj): obj is TEngine.Queryables { return !obj || (obj as TEngine.Queryables).$r$ }
export function isDeferred(obj): obj is TEngine.Deferred { return obj && (obj as TEngine.Deferred).$d$ }
export function isTemporary(obj): obj is TEngine.TempProc { return obj && (obj as TEngine.TempProc).$t$ }
export function isToAtomize(obj): obj is TEngine.ToAtomize { return obj && !(obj as TEngine.Queryables).$r$ && !(obj as TEngine.Deferred).$d$ && !(obj as TEngine.TempProc).$t$ }

//*******************************************************
//        makeTemporary
export const makeTemporary = (proc: TEngine.TempProc) => {
    proc.$t$ = true
    return proc as any
}

//*******************************************************
//        processTree
export const processTree = (
    value: TEngine.Ruleset | TEngine.Ruleset[], atomizedVariants: TEngine.QueryableItems,
    path: string, pseudoPrefixes: string[], conditions: TEngine.Conditions
) => {
    const items = isToAtomizeArray(value) ? value : [value]
    const indexToPath = (idx: number) => items.length <= 1 ? '' : `[${idx}]`
    items.forEach((it, idx) => {
        if (!it) return
        if (isTemporary(it)) {
            it(atomizedVariants, path + indexToPath(idx), pseudoPrefixes, conditions)
        } else if (isDeferred(it)) {
            atomizedVariants.push(it)
        } else if (isAtomized(it)) {
            if (pseudoPrefixes && pseudoPrefixes.length > 0)
                throw 'Web pseudo properties cannot contain already atomized value'
            it.forEach(it => {
                if (!conditions || conditions.length === 0) //... and no conditions => done
                    pushToAtomizedVariants(atomizedVariants, it, null)
                else { // ... and conditions => merge conditions
                    pushToAtomizedVariants(atomizedVariants, [...it],
                        it.conditions ? [...conditions, ...it.conditions] : conditions)
                }
            })
        } else { // it: toAtomize
            processPseudosAndAtomize(
                it,
                atomizedVariants, path + indexToPath(idx),
                pseudoPrefixes, conditions
            )
        }
    })
}

//*********************************************************
//  PRIVATE
//*********************************************************

//*******************************************************
//        processAllInnerVariants

// recursivelly process web pseudos (:hover, :active) and its queryable parts

const processPseudosAndAtomize = (
    ruleset: TEngine.ToAtomize,
    atomizedVariants: TEngine.QueryableItems, path: string, pseudoPrefixes: string[], conditions: TEngine.Conditions,
) => {

    const inner: TEngine.QueryableItems = []

    // process and remove conditional parts of self and web pseudo (e.g. :hover etc.)
    // push them to 'inner'
    for (const p in ruleset) {
        const value = ruleset[p] as TEngine.ToAtomize
        if (!value || p.charAt(0) === '$' || (typeof value !== 'object' && !isTemporary(value))) continue
        processTree(value, inner, `${path}/${p}`, [...pseudoPrefixes, p], conditions)
        delete ruleset[p]
    }

    // atomize pure ruleset tree with pseudo (and without all conditions, deffer's etc.),
    // then push to result
    atomizeNonVariant(atomizedVariants, wrapPseudoPrefixes(ruleset, pseudoPrefixes), conditions, path)

    // push 'inner' to result
    if (inner.length > 0)
        Array.prototype.push.apply(atomizedVariants, inner)

}

// *********** utils

const pushToAtomizedVariants = (
    atomizedVariants: TEngine.QueryableItems, 
    variant: TEngine.Queryable, conditions: TEngine.Conditions
    ) => {
    if (!variant) return
    if (conditions && conditions.length > 0) variant.conditions = conditions
    atomizedVariants.push(variant)
}

const atomizeNonVariant = (
    atomizedVariants: TEngine.QueryableItems, sourceRuleset: TEngine.Rulesets,
    conditions: TEngine.Conditions, path: string
) => {
    if (!sourceRuleset) return
    if (isDeferred(sourceRuleset)) {
        atomizedVariants.push(sourceRuleset)
        return
    }
    // platform specific atomize: e.g. for WEB convert ruleset to list of atomized CSS classes
    const variant = platform.toPlatformAtomizeRuleset(sourceRuleset, path)
    pushToAtomizedVariants(atomizedVariants, variant, conditions)
}

