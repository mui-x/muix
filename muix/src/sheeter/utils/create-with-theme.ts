import warning from 'warning'
import { isTemporary } from './atomize-low'

export const createWithTheme = (valueOrCreator: ((theme) => any) | any, theme) => {
  if (typeof valueOrCreator === 'function' && !isTemporary(valueOrCreator))
    // apply creator to theme:
    return valueOrCreator(theme)
  else // return value
    return valueOrCreator
}