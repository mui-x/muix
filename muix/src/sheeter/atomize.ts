import warning from 'warning';

import { TAtomize, TSheeter, TComponents, TCommonStyles } from 'reactxx-typings'
import { createWithTheme } from './utils/create-with-theme'
import { adjustAtomizedLow } from './atomize-low'

// muttable
export const atomizeSheet = <R extends TSheeter.Shape = TSheeter.Shape>(sheet: TSheeter.PartialSheetOrCreator<R>, theme?, path: string = 'sheet') => {
    if (!sheet) return null
    const sh: TSheeter.Sheet = createWithTheme(sheet, theme)
    for (const p in sh) {
        const at = atomizeRuleset(sh[p], theme, (path ? path + '.' : '') + p)
        if (at) sh[p] = at
        else delete sh[p]
    }
    return sh as any as TAtomize.Sheet<R>
}

// muttable
export const atomizeRuleset = <T extends TCommonStyles.RulesetNativeIds = 'Text', R extends TSheeter.Shape = TSheeter.Shape>(
    ruleset: TSheeter.RulesetOrAtomizedCreator<T, R>, theme?, path: string = '.'
) => {
    if (!ruleset) return null

    const rs = createWithTheme(ruleset, theme) as TAtomize.Source

    if (!rs) return null

    const list = wrapRuleset([])
    adjustAtomizedLow(list, rs, path, [], [])

    return list.length === 0 ? null : list
}

export const wrapRuleset = (ruleset) => {
    (ruleset as TAtomize.Ruleset).$r$ = true
    if (window.__TRACE__)
        ruleset['toJSON'] = toJSON.bind(ruleset)
    return ruleset as TAtomize.Ruleset
}
function toJSON_() {
    return (this as TAtomize.Ruleset).map(v => v.conditions ? {
        conditions: v.conditions,
        items: v
    } : v)
}
function toJSON() {
    return (this as TAtomize.Ruleset).map(v => v.conditions ? [...v.conditions, ...v] : v)
}

// muttable (at least for native)
export const atomizeStyle = (style: TSheeter.StyleOrCreator, theme?, path: string = '.') => {
    if (!style) return null
    if (window.isWeb)
        return createWithTheme(style, theme) as TSheeter.StyleOrAtomizedWeb
    else
        return atomizeRuleset(style as TSheeter.RulesetOrAtomized, theme, path)
}

export function isReactXXComponent(obj): obj is TComponents.ComponentType {
    return (obj as TComponents.ComponentType).$c$
}