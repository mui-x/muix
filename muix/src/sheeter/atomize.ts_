import warning from 'warning';

import { TAtomize, TSheeter, TComponents, TVariants } from 'reactxx-typings'
import { createWithTheme } from './utils/create-with-theme'
import { atomizeRulesetLow } from './atomize-low'

export const atomizeSheet = <R extends TSheeter.Shape = TSheeter.Shape>(sheet: TSheeter.SheetOrCreator<R>, theme?, path: string = 'sheet') => {
    if (!sheet) return null
    const sh: TSheeter.Sheet = createWithTheme(sheet, theme, window.__TRACE__)
    for (const p in sh) {
        const at = atomizeRuleset(sh[p], p, theme, (path ? path + '.' : '') + p)
        if (at) sh[p] = at
        else delete sh[p]
    }
    return sh as any as TAtomize.Sheet<R>
}

export const atomizeRuleset = (ruleset: TSheeter.ClassNameOrCreator, rulesetName: string, theme?, path: string = '.') => {
    if (!ruleset) return null
    const rs: TSheeter.RulesetOrAtomized = createWithTheme(ruleset, theme)

    if (!rs) return null

    const name = rulesetName || rs['name'] || '.'

    const list: TAtomize.Variants = []
    atomizeRulesetLow(rs, list, path, [], [])

    return list.length === 0 ? null : {
        name,
        list,
        [TAtomize.TypedInterfaceTypes.prop]: TAtomize.TypedInterfaceTypes.atomizedRuleset
    } as TAtomize.AtomizedRuleset


    //return atomizeRulesetLow(rs, rulesetName)
}

export const atomizeStyle = (style: TSheeter.StyleOrCreator, theme?, path: string = '.') => {
    if (!style) return null
    if (window.isWeb)
        return createWithTheme(style, theme) as TSheeter.StyleOrAtomizedWeb
    else
        return atomizeRuleset(style as TSheeter.ClassNameOrCreator, '', theme, path)
}

export function isAtomizedRuleset(obj: Object): obj is TAtomize.AtomizedRuleset {
    return obj && obj['~'] === TAtomize.TypedInterfaceTypes.atomizedRuleset
}

export function isAtomicArray(obj): obj is TAtomize.AtomicArray {
    return obj && obj[TAtomize.TypedInterfaceTypes.prop] === TAtomize.TypedInterfaceTypes.atomicArray
}
export function isAtomizedSheet<R extends TSheeter.Shape = TSheeter.Shape>(sheet: TSheeter.SheetX): sheet is TAtomize.Sheet<R> {
    for (const p in sheet)
        return isAtomizedRuleset(sheet[p])
    return true
}

export function isReactXXComponent(obj): obj is TComponents.ComponentType {
    return obj[TAtomize.TypedInterfaceTypes.prop] === TAtomize.TypedInterfaceTypes.reactxxComponent
}

