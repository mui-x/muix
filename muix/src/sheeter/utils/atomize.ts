import { TEngine } from 'reactxx-typings';
import { adjustAtomizedLow, isDeferred } from './atomize-low';
import { createWithTheme } from './create-with-theme';



// muttable
export const atomizeSheet = (sheet: TEngine.SheetOrCreator, theme?, path: string = 'sheet') => {
    if (!sheet) return null
    const sh: TEngine.Sheet = createWithTheme(sheet, theme)
    for (const p in sh) {
        const at = atomizeRuleset(sh[p], theme, (path ? path + '.' : '') + p)
        if (at) sh[p] = at
        else delete sh[p]
    }
    return sh
}

// muttable
export const atomizeRuleset = (
    ruleset: TEngine.RulesetOrCreator, theme?, path: string = ''
) => {
    if (!ruleset) return null

    const rs = createWithTheme(ruleset, theme)

    if (!rs) return null

    const list = wrapRuleset([])
    adjustAtomizedLow(rs, list, path, [], [])

    return list.length === 0 ? null : list
}

export const wrapRuleset = ruleset => {
    (ruleset as TEngine.Queryables).$r$ = true
    if (window.__TRACE__)
        ruleset['toJSON'] = toJSON.bind(ruleset)
    return ruleset as TEngine.Queryables
}
function toJSON() {
    return (this as TEngine.Queryables).map(v => {
        if (isDeferred(v)) return 'DEFFERED'
        return v
    })
}

// muttable (at least for native)
export const atomizeStyle = (style: TEngine.Style, theme?) => {
    if (!style) return null

    style = createWithTheme(style, theme)

    const { $web, $native } = style

    if ($web || $native) style = { ...style }

    if (window.isWeb && $web)
        Object.assign(style, $web)
    else if (!window.isWeb && $native)
        Object.assign(style, $native)

    delete style.$web
    delete style.$native
    
    return style
}

