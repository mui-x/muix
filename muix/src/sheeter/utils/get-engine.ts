import { TTyped } from 'reactxx-typings';
import { $width, $web, $native, $hot, $if, $ifelse } from '../queryable/index'
import { toClassNamesWithQuery } from '../utils/to-classnames'
import { mergeRulesets, mergeSheets } from '../utils/merge'
import { atomizeRuleset, atomizeSheet } from '../utils/atomize'

export const getEngine = <R extends TTyped.Shape>() => untypedEngine as TTyped.TypedEngine<R>

const untypedEngine = {
    WEB: $web,
    NATIVE: $native,
    HOT: $hot,
    IF: $if,
    IFELSE: $ifelse,
    WIDTH: $width,
    THEMED: p => p,
    STYLE: (...p) => p,
    $atomizeSheet: atomizeSheet as any,
    $mergeSheets: mergeSheets as any,
    $atomizeRuleset: atomizeRuleset as any,
    ATOMIZE: (...pars: any[]) => atomizeRuleset(pars, null, '$atomize'),
    $mergeRulesets: mergeRulesets as any,
    $toClassNames: toClassNamesWithQuery as any
} as TTyped.TypedEngine<TTyped.Shape>
