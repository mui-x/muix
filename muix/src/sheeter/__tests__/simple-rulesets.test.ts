import { $W, $T, $V, $I, V, T, I, TTyped } from 'reactxx-typings'
import { getEngine } from '../utils/get-engine'
import {atomizeRuleset} from '../utils/atomize'

import { theme, Theme } from "reactxx-typings-test/shape.t"
import { initPlatform } from "./init-platform.t"

interface Shape {
  theme: Theme
  sheetQuery: {opened: boolean}  
}

const { WEB, HOT, NATIVE, ATOMIZE
} = getEngine<Shape>()

const t = WEB({ color: 'green' })

describe("SHEETER SIMPLE RULESET", () => {

  const doTest = (isWeb: boolean) => {
    beforeEach(() => initPlatform(isWeb))
    let ruleset

    it("01 (null)", () => {
      ruleset = ATOMIZE(null)
      expect(ruleset).toMatchSnapshot()
    })

    it("02 ({})", () => {
      ruleset = ATOMIZE({})
      expect(ruleset).toMatchSnapshot()
    })

    it("03 ([null, undefined, {}])", () => {
      ruleset = ATOMIZE(null, undefined, {})
      expect(ruleset).toMatchSnapshot()
    })

    it("04 ({ color: 'red', margin: 0 })", () => {
      ruleset = ATOMIZE({ color: 'red', margin: 0 })
      expect(ruleset).toMatchSnapshot()
    })

    it("05 ([{ color: 'red' }, { margin: 0 }])", () => {
      ruleset = ATOMIZE({ color: 'red' }, { margin: 0 })
      expect(ruleset).toMatchSnapshot()
    })

    it("06 (...$web: { color: 'green' }, $native: { color: 'blue' })", () => {
      ruleset = ATOMIZE<T>(
        {
          color: 'red',
        },
        WEB({ color: 'green' }),
        NATIVE<$T>({ color: 'blue' }),
        NATIVE({ color: 'blue' }),
      )
      expect(ruleset).toMatchSnapshot()
    })

    it("07 atomizeRuleset(atomizeRuleset({color: 'red'", () => {
      ruleset = ATOMIZE<T>(ATOMIZE<T>({
        color: 'red'
      }))
      expect(ruleset).toMatchSnapshot()
    })

    it("08 atomizeRuleset({$web: atomizeRuleset({ :hover color: 'red', $native: atomizeRuleset({ color: 'green'", () => {
      ruleset = ATOMIZE<T>(
        WEB(ATOMIZE<$W>({ ':hover': { color: 'red' } })),
        NATIVE<$T>(ATOMIZE<T>({ color: 'green' }))
      )
      expect(ruleset).toMatchSnapshot()
    })

    it("09 ERROR for atomizeRuleset({:hover': atomizeRuleset(...", () => {
      if (!window.isWeb) return
      const fnc = () => {
        ATOMIZE<V>(
          WEB({
            ':hover': ATOMIZE<$W>(
              { color: 'red' }
            ),
          }))
      }
      expect(fnc).toThrow(/.*/)
    })

    it("10 ERROR for atomizeRuleset({:hover': [ARRAY]", () => {
      if (!window.isWeb) return
      const fnc = () => ATOMIZE<V>(
        WEB({
          ':hover': [
            {},
            ATOMIZE<$W>({ color: 'red' }),
            { margin: 10 }
          ]
        })
      )
      expect(fnc).toThrow(/.*/)
    })

    it("11 with theme", () => {
      ruleset = atomizeRuleset((theme: any) => [
        theme.primary.normal,
        WEB(theme.secondary.normal),
        NATIVE<$T>(theme.secondary.normal),
      ], theme)
      expect(ruleset).toMatchSnapshot()
    })

    it("12 ATOMIZE modifies source object", () => {
      const source = [
        { color: 'blue' },
        WEB({ color: 'red' }),
        NATIVE({ color: 'green' }),
      ]
      ruleset = ATOMIZE(...source)
      expect(ruleset).toMatchSnapshot()
      expect(source).toMatchSnapshot()
    })

    it("13 condition in pseudo", () => {
      if (!window.isWeb) return
      ruleset = ATOMIZE<V>(
        WEB({
          ':hover': WEB({ color: 'red' }),
        }),
      )
      expect(ruleset).toMatchSnapshot()
    })

    it("14 mix in pseudo array", () => {
      if (!window.isWeb) return
      ruleset = ATOMIZE(
        WEB({
          color: 'green',
          ':hover': [
            { color: 'blue' },
            WEB({ ':active': { color: 'red' } })
          ],
        }),
      )
      expect(ruleset).toMatchSnapshot() 
    })

    it("15 just $hot", () => {
      ruleset = ATOMIZE<T>(
        { color: 'green' },
        HOT<T>(state => ({
          color: 'red'
        })),
      )
      expect(ruleset).toMatchSnapshot()
    })

  }

  describe("## NATIVE ##", () => doTest(false))
  describe("## WEB ##", () => doTest(true))

})

