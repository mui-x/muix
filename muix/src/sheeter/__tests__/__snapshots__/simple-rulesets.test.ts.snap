// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SHEETER SIMPLE RULESET ## NATIVE ## 01 (null) 1`] = `null`;

exports[`SHEETER SIMPLE RULESET ## NATIVE ## 02 ({}) 1`] = `null`;

exports[`SHEETER SIMPLE RULESET ## NATIVE ## 03 ([null, undefined, {}]) 1`] = `null`;

exports[`SHEETER SIMPLE RULESET ## NATIVE ## 04 ({ color: 'red', margin: 0 }) 1`] = `
Array [
  Object {
    "color": "red @",
    "margin": "0 @",
  },
]
`;

exports[`SHEETER SIMPLE RULESET ## NATIVE ## 05 ([{ color: 'red' }, { margin: 0 }]) 1`] = `
Array [
  Object {
    "color": "red @[0]",
  },
  Object {
    "margin": "0 @[1]",
  },
]
`;

exports[`SHEETER SIMPLE RULESET ## NATIVE ## 06 (...$web: { color: 'green' }, $native: { color: 'blue' }) 1`] = `
Array [
  Object {
    "color": "red @[0]",
  },
  Object {
    "color": "blue @[2]/$native",
  },
]
`;

exports[`SHEETER SIMPLE RULESET ## NATIVE ## 07 atomizeRuleset(atomizeRuleset({color: 'red' 1`] = `
Array [
  Object {
    "color": "red @",
  },
]
`;

exports[`SHEETER SIMPLE RULESET ## NATIVE ## 08 atomizeRuleset({$web: atomizeRuleset({ :hover color: 'red', $native: atomizeRuleset({ color: 'green' 1`] = `
Array [
  Object {
    "color": "green @",
  },
]
`;

exports[`SHEETER SIMPLE RULESET ## NATIVE ## 11 with theme 1`] = `
Array [
  Object {
    "backgroundColor": "darkgreen @[0]",
    "color": "white @[0]",
  },
  Object {
    "backgroundColor": "lightblue @[2]/$native",
    "color": "black @[2]/$native",
  },
]
`;

exports[`SHEETER SIMPLE RULESET ## NATIVE ## 12 ATOMIZE modifies source object 1`] = `
Array [
  Object {
    "color": "blue @[0]",
  },
  Object {
    "color": "green @[2]/$native",
  },
]
`;

exports[`SHEETER SIMPLE RULESET ## NATIVE ## 12 ATOMIZE modifies source object 2`] = `
Array [
  Object {
    "color": "blue",
  },
  false,
  [Function],
]
`;

exports[`SHEETER SIMPLE RULESET ## NATIVE ## 15 just $hot 1`] = `
Array [
  Object {
    "color": "green @[0]",
  },
  "DEFFERED",
]
`;

exports[`SHEETER SIMPLE RULESET ## WEB ## 01 (null) 1`] = `null`;

exports[`SHEETER SIMPLE RULESET ## WEB ## 02 ({}) 1`] = `null`;

exports[`SHEETER SIMPLE RULESET ## WEB ## 03 ([null, undefined, {}]) 1`] = `null`;

exports[`SHEETER SIMPLE RULESET ## WEB ## 04 ({ color: 'red', margin: 0 }) 1`] = `
Array [
  Array [
    ".a { color:red } @",
    ".b { margin:0 } @",
  ],
]
`;

exports[`SHEETER SIMPLE RULESET ## WEB ## 05 ([{ color: 'red' }, { margin: 0 }]) 1`] = `
Array [
  Array [
    ".a { color:red } @[0]",
  ],
  Array [
    ".b { margin:0 } @[1]",
  ],
]
`;

exports[`SHEETER SIMPLE RULESET ## WEB ## 06 (...$web: { color: 'green' }, $native: { color: 'blue' }) 1`] = `
Array [
  Array [
    ".a { color:red } @[0]",
  ],
  Array [
    ".b { color:green } @[1]/$web",
  ],
]
`;

exports[`SHEETER SIMPLE RULESET ## WEB ## 07 atomizeRuleset(atomizeRuleset({color: 'red' 1`] = `
Array [
  Array [
    ".a { color:red } @",
  ],
]
`;

exports[`SHEETER SIMPLE RULESET ## WEB ## 08 atomizeRuleset({$web: atomizeRuleset({ :hover color: 'red', $native: atomizeRuleset({ color: 'green' 1`] = `
Array [
  Array [
    ".a:hover { color:red } @/:hover",
  ],
]
`;

exports[`SHEETER SIMPLE RULESET ## WEB ## 11 with theme 1`] = `
Array [
  Array [
    ".a { color:white } @[0]",
    ".b { background-color:darkgreen } @[0]",
  ],
  Array [
    ".c { color:black } @[1]/$web",
    ".d { background-color:lightblue } @[1]/$web",
  ],
]
`;

exports[`SHEETER SIMPLE RULESET ## WEB ## 12 ATOMIZE modifies source object 1`] = `
Array [
  Array [
    ".a { color:blue } @[0]",
  ],
  Array [
    ".b { color:red } @[1]/$web",
  ],
]
`;

exports[`SHEETER SIMPLE RULESET ## WEB ## 12 ATOMIZE modifies source object 2`] = `
Array [
  Object {
    "color": "blue",
  },
  [Function],
  false,
]
`;

exports[`SHEETER SIMPLE RULESET ## WEB ## 13 condition in pseudo 1`] = `
Array [
  Array [
    ".a:hover { color:red } @/$web/:hover/$web",
  ],
]
`;

exports[`SHEETER SIMPLE RULESET ## WEB ## 14 mix in pseudo array 1`] = `
Array [
  Array [
    ".c { color:green } @/$web",
  ],
  Array [
    ".a:hover { color:blue } @/$web/:hover[0]",
  ],
  Array [
    ".b:hover:active { color:red } @/$web/:hover[1]/$web/:active",
  ],
]
`;

exports[`SHEETER SIMPLE RULESET ## WEB ## 15 just $hot 1`] = `
Array [
  Array [
    ".a { color:green } @[0]",
  ],
  "DEFFERED",
]
`;
